File: src/main/java/net/nodus/database/common/BaseDocument.java
package net.nodus.database.common;

import java.time.Instant;
import lombok.Getter;
import lombok.Setter;
import org.springframework.data.annotation.CreatedDate;
import org.springframework.data.mongodb.core.index.Indexed;

@Getter
@Setter
public abstract class BaseDocument {

    @CreatedDate
    private Instant createdAt = Instant.now();

    @Indexed
    private Instant deletedAt;

    public void softDelete() {
        softDelete(Instant.now());
    }

    public void softDelete(Instant now) {
        this.deletedAt = now;
    }

    public void restore() {
        restore(Instant.now());
    }

    public void restore(Instant now) {
        this.deletedAt = null;
    }

    public boolean isDeleted() {
        return deletedAt != null;
    }
}

File: src/main/kotlin/net/nodus/common/BaseDocument.kt
package net.nodus.common

import org.springframework.data.annotation.CreatedDate
import org.springframework.data.mongodb.core.index.Indexed
import java.time.Instant

abstract class BaseDocument (
    @CreatedDate
    var createdAt: Instant = Instant.now(),

    @Indexed
    var deletedAt: Instant? = null,
) {
    open fun softDelete(now: Instant = Instant.now()) {
        deletedAt = now
    }

    open fun restore(now: Instant = Instant.now()) {
        deletedAt = null
    }

    val isDeleted: Boolean
        get() = deletedAt != null
}